packages/compiler/core/helpers/usage-resolver.ts

```diff
@@ -0,0 +1,130 @@
+import {
+  EnumType,
+  InterfaceType,
+  ModelType,
+  NamespaceType,
+  OperationType,
+  TupleType,
+  Type,
+  UnionType,
+} from "../types.js";
+
+// prettier-ignore
+export enum UsageFlags {
+  None   = 0,
+  Input  = 1 << 1,
+  Output = 1 << 2,
+}
+
+export type TrackableType = ModelType | EnumType | UnionType | TupleType;
+
+export interface UsageTracker {
+  readonly types: readonly TrackableType[];
+  isUsedAs(type: TrackableType, usage: UsageFlags): boolean;
+}
+
+/**
+ * Resolve usage(input, output or both) of various types in the given namespace.
+ * Will recursively scan all namespace, interfaces and operations contained inside the namespace.
+ * @param namespace Entrypoint namespace to get usage from.
+ * @returns Map of types to usage.
+ */
+export function resolveUsages(namespace: NamespaceType): UsageTracker {
```

Comment:
Based on the issue, it feels like we need to provide a specific exported function that will return a list of models that are used in a specific operation (the input/output flags are nice), in addition to this function which answers the question over all operations in the namespace.